contracts/BoostedMasterChefJoe.sol

```diff
@@ -0,0 +1,621 @@
+    /// the pending amount when a user claims.
+    mapping(uint256 => mapping(address => uint256)) public claimableJoe;
+
+    event Add(uint256 indexed pid, uint256 allocPoint, IERC20 indexed lpToken, IRewarder indexed rewarder);
+    event Set(uint256 indexed pid, uint256 allocPoint, IRewarder indexed rewarder, bool overwrite);
+    event Deposit(address indexed user, uint256 indexed pid, uint256 amount);
+    event Withdraw(address indexed user, uint256 indexed pid, uint256 amount);
+    event UpdatePool(uint256 indexed pid, uint256 lastRewardTimestamp, uint256 lpSupply, uint256 accJoePerShare);
+    event Harvest(address indexed user, uint256 indexed pid, uint256 amount);
+    event EmergencyWithdraw(address indexed user, uint256 indexed pid, uint256 amount);
+    event Init();
+    event MaxBoostFactorChanged(uint256 maxBoostFactor);
+
+    /// @param _MASTER_CHEF_V2 The JoeSwap MCJV2 contract address.
+    /// @param _joe The JOE token contract address.
+    /// @param _MASTER_PID The pool ID of the dummy token on the base MCJV2 contract.
+    constructor(
+        IMasterChef _MASTER_CHEF_V2,
+        IERC20 _joe,
+        IERC20 _veJoe,
+        uint256 _MASTER_PID
+    ) public {
+        MASTER_CHEF_V2 = _MASTER_CHEF_V2;
+        JOE = _joe;
+        VEJOE = _veJoe;
+        MASTER_PID = _MASTER_PID;
+    }
+
+    /// @notice Deposits a dummy token to `MASTER_CHEF_V2` MCJV2. This is required because MCJV2 holds the minting rights for JOE.
+    /// Any balance of transaction sender in `dummyToken` is transferred.
+    /// The allocation point for the pool on MCJV2 is the total allocation point for all pools that receive double incentives.
+    /// @param dummyToken The address of the ERC-20 token to deposit into MCJV2.
+    function init(IERC20 dummyToken) external onlyOwner {
+        uint256 balance = dummyToken.balanceOf(msg.sender);
+        require(balance != 0, "MasterChefV2: Balance must exceed 0");
+        dummyToken.safeTransferFrom(msg.sender, address(this), balance);
+        dummyToken.approve(address(MASTER_CHEF_V2), balance);
+        MASTER_CHEF_V2.deposit(MASTER_PID, balance);
+        emit Init();
+    }
+
+    /// @notice Returns the number of BMCJ pools.
+    function poolLength() external view returns (uint256 pools) {
+        pools = poolInfo.length;
+    }
+
+    /// @notice Add a new LP to the pool. Can only be called by the owner.
+    /// DO NOT add the same LP token more than once. Rewards will be messed up if you do.
+    /// @param allocPoint AP of the new pool.
+    /// @param _lpToken Address of the LP ERC-20 token.
+    /// @param _rewarder Address of the rewarder delegate.
+    function add(
+        uint256 allocPoint,
+        IERC20 _lpToken,
+        IRewarder _rewarder
+    ) external onlyOwner {
+        require(!lpTokens.contains(address(_lpToken)), "add: LP already added");
+        require(poolInfo.length < 50, "Too many pools");
```

Comment:
do we want to make the max number of pools configurable?

Reply:
We are limited by gas here and ATM 50 is just a figure I was told, I need to do proper testing to find out a real max number, after that not sure there is any point in making it configurable.